lib/CL/devices/proxy/pocl_proxy.cpp

```diff
@@ -1277,55 +1328,119 @@ pocl_proxy_setup_metadata (cl_device_id device, cl_program program,
+  if (err) {
+    POCL_MSG_ERR("proxy metadata setup error 2: %i", err);
+    return 0;
+  }
+
+  if (d->backend->supports_il && program->program_il_size > 0) {
+
+    SPIRVParser::OpenCLFunctionInfoMap infoMap;
+    bool parseRet = SPIRVParser::parseSPIRV((const int32_t *)program->program_il,
+                               program->program_il_size / 4, infoMap);
+
+    if (!parseRet) {
+      POCL_MSG_WARN("Proxy could not parse SPIR-V metadata.\n");
+      return 0;
     }
-  else
+    assert(infoMap.size() == num_kernels);
+
+    char string_value[POCL_MAX_PATHNAME_LENGTH];
+    int index = 0;
+    for (auto &argInfo : infoMap) {
+      SPIRVParser::mapToPoCLMetadata(argInfo, program->num_devices, &p[index]);
+
+      err = get_kernel_info(&p[index], d->device_id, kernels[index]);
+      if (err != CL_SUCCESS) {
+        POCL_MSG_WARN("Proxy could not get kernel info of SPIR-V kernel.\n");
+        return 0;
+      }
+
+      // these checks are there to make sure that the results match
+
+      err = clGetKernelInfo(kernels[index], CL_KERNEL_FUNCTION_NAME,
+                            POCL_MAX_PATHNAME_LENGTH, string_value, NULL);
+      assert(strcmp(p[index].name, string_value) == 0);
+
+      cl_uint prox_num_args;
+      err = clGetKernelInfo(kernels[index], CL_KERNEL_NUM_ARGS,
+                            sizeof(prox_num_args), &prox_num_args, NULL);
+      assert(err == CL_SUCCESS);
+      assert(p[index].num_args = prox_num_args);
+
+      err = clReleaseKernel(kernels[index]);
+      assert(err == CL_SUCCESS);
+      index++;
+    }
+
+  } else {
+    for (cl_uint i = 0; i < num_kernels; ++i) {
+      err = get_kernel_metadata(p + i, program->num_devices, proxy_prog,
+                                d->device_id, kernels[i]);
+      if (err != CL_SUCCESS) {
+        POCL_MSG_WARN("Failed to kernel metadata for index %d.\n", i);
+        return 0;
+      }
+
+      err = clReleaseKernel(kernels[i]);
+      assert(err == CL_SUCCESS);
+    }
+  }
+
+  program->kernel_meta = p;
+
+  POCL_MSG_PRINT_PROXY("Num kernels: %zu\n", program->num_kernels);
+  return 1;
+}
+
+int pocl_proxy_supports_binary(cl_device_id device, size_t length,
+                               const char *binary) {
+  // Check that the proxy device supports SPIR-V, see link for more details
+  // on the SPIR-V extension:
+  // https://registry.khronos.org/OpenCL/sdk/3.0/docs/man/html/cl_khr_il_program.html
+  cl_int err;
+  proxy_device_data_t *d = (proxy_device_data_t *)device->data;
+  size_t extension_length = 0;
+  err = clGetDeviceInfo(d->device_id, CL_DEVICE_EXTENSIONS, 0, NULL,
+                        &extension_length);
+  if (err != CL_SUCCESS) {
     return 0;
+  }
+  char *extensions = (char *)malloc(extension_length);
+  err = clGetDeviceInfo(d->device_id, CL_DEVICE_EXTENSIONS, extension_length,
+                        extensions, NULL);
+  int extension_support = 1;
+  if (err != CL_SUCCESS || strstr(extensions, "cl_khr_il_program") == NULL)
```

Comment:
Small nitpick: AFAIK the presence of this extension means the platform supports `clCreateProgramWithILKHR`; however, from OpenCL 2.1 this API is "core" meaning it should be always available as `clCreateProgramWithIL` (irrespectively if the platform reports support for `cl_khr_il_program` extension)

Reply:
I just checked with example0 and an Nvidia gpu which doesn't have the `cl_khr_il_program` extension and it does run spir-v